### PR TITLE
Feature: character search/filter

### PR DESCRIPTION
**Pull request by adamsilverstein**
## Description
Add a search input that filters on a basic string search that matches from the beginning of string (also tried substring)

Fixes https://github.com/Dayjo/react-character-map/issues/4.

## Changes
* Move `categoryList` and `charList` into state, calculating them once during `componentDidMount`. Also store the full character list in state.
* Add a helper `charListFromCharacters` to build up the Character Map display (uses existing code). This helper is reused to generate the display for the search filtered
* Add a Filter `input` that initiates a search - filtering the character list based on a match at the beginning of the "name" when a name is available. 
* Hide the category tabs when filtering
* Restore the full original character map when the search field is cleared

## Screencast
Looks something like this in our upcoming plugin (shown here with substring match vs start of string patch):

-----
Note: A substring result might make finding things easier, i tried both matching at the beginning of the name and a substring, something like chess found the chess pieces even though the name is something like "White Chess King". Changing this === 0 to !== -1 will do that: https://github.com/Dayjo/react-character-map/compare/master...adamsilverstein:feature/searchplus?expand=1#diff-072f43aab5d92c8150d8860cf4349128R50

### CODE REVIEW

File: src/component/CharacterMap.js [hunk 1]
```diff
@@ -27,21 +32,48 @@ class CharacterMap extends React.Component {
         return this.props.onSelect(char, e.target);
     }
 
-    render() {
+    // Filter the displayed characters.
+    handleSearchChange( e ) {
+        const search = e.target.value;
+        const {fullCharList,charList} = this.state;
+        console.log( search );
```
Comment (Dayjo) [inline src/component/CharacterMap.js]:
```suggestion

```

**Comment by Dayjo** [comment]
Looking great man, and thanks for not just loading in lodash or something and doing it vanillary 👍.

Should be getting my laptop back tonight so I'll try and get these changes merged in in the next few days!

File: src/component/CharacterMap.js (hunk 1 above)
Comment (adamsilverstein) [reply src/component/CharacterMap.js]: :) yep!

**Comment by adamsilverstein** [comment]
Absolutely, glad you like it! I have tried to stick to your coding standards generally as well.

Thinking about the search approach, I have an idea about improving the search to combine the 'starts with' and 'substring' type of searching... We can use the 'starts with' search for a single character, then substring for 2 or more and also sort the substring matches by position, so those that matched the beginning of the string would be at the top of the list. 

I'll give this a try and update the PR if I feel it improves results.

**Comment by Dayjo** [comment]
"coding standards" 😆

**Comment by adamsilverstein** [comment]
Couple of improvements:
* moved relatively expensive searching to its own helper function and added caching so searches only happen once per search string
* If search string is one character long, look for names that start with that character.
* When the search string is two or more characters, do a full search of the name. As they are collected, results are sorted so matches closest to front of name show first.

**Pull request merged**